python.md

```diff
@@ -637,6 +638,30 @@ Why?
 2. It also avoids issues where Celery tasks are publishing on one day but get
    executed on another. It removes an assumption from the code. 
 
+### <a name="time-periods">Modelling periods of time</a>
+
+It's common for domain objects to model some period of time that defines when an
+object is "active" or "valid". When faced with this challenge, prefer to use
+_datetime_ fields where the upper bound is nullable and exclusive. Eg:
```

Comment:
I think the concept of a datetime being _exclusive_ is a bit slippery. Do you think of `2019-01-01 12:00:00` as representing a zero length instant at noon or the period of time between `12:00:00` and `12:00:01`?

If you think of datetimes as representing a zero length instant, I would argue that there is no difference between an exclusive upper bound and an inclusive one.

If you think of datetimes as representing a small but non-zero length of time (A second, a microsecond,  whatever) then exclusive/inclusive has a meaning.

Reply:
@richardcooper yes, it's domain-specific, and nothing to do with datetimes *per se*, but rather to do with the fact that in our industry, there's no such thing as a one-second electricity agreement, whereas If we were dealing with network packets being `sent_at` and `received_at`, then we wouldn't be saying that one which was `sent_at` 00:00:00 and also `received_at` 00:00:00 had a 'zero-length' journey through the network: although there might be an equivalent 'meaningless' level of precision which we could use (picoseconds?).

At the other end of the scale, if you knew that supply periods had a minimum period of a week, then you could model zero-length supply periods with dates, saying that if the supply start and end dates are the same, then it's a zero-length supply period. So I think this is domain-specific rather than instrinsic to the type of object we're using. 

Still, so long as you understand that, it does seem to be a useful convention to be able to make use of. Maybe it'll come back and bite us when we move to a future of being able to switch agreements ten times within a day, like you can with insurance I hear.

Reply:
> I think the concept of a datetime being _exclusive_ is a bit slippery. Do you think of `2019-01-01 12:00:00` as representing a zero length instant at noon or the period of time between `12:00:00` and `12:00:01`?

The former.

> If you think of datetimes as representing a zero length instant, I would argue that there is no difference between an exclusive upper bound and an inclusive one.

I dunno - it's still a point on a line that you can test equality for. It matters for SQL queries where you want to say:
```sql
...WHERE dt >= available_from and dt < available_to
```
There's a big difference there between `dt < available_to` and `dt <= available_to`, especially when dealing with lots of midnight timestamps (as we do).

But perhaps this is just semantics. What's a better way of saying to that you shouldn't test equality on the upper bound?